### PR TITLE
ASRC-328: Vitals defect fix.

### PR DESCRIPTION
**Pull request by jamesbigmac**
Did my best to fix a defect with how vitals were retrieved. Still awaiting more information from Jeff because right now the code is guessing at how vitals are returned.

### CODE REVIEW

File: srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java [hunk 1]
```diff
@@ -223,25 +225,69 @@ private void parseRecentVitalResults(final Patient patient, final List<String> v
...
-        patient.setBmi(new RetrievedValue(
-            Double.parseDouble(bmiLineTokens.get(bmiLineTokens.size()-2)),
-            patient.getWeight().getMeasureDate(),
-            ""));
+        for(final String line: vitalResults)
+        {
+            // For each line, determine if we need to pull the vital information for that line
+            // Each entry comes with an optional date and time.
+            final Pattern pattern = Pattern.compile(
+                    "(.+:)\\s+" // Match to the first colon in order to get the name of the vital
```
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java]: I like how you split up the regex with comments like this. It's a good way of documenting it.

Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java]: Why capture the colon?

File: srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java [hunk 2]
```diff
@@ -223,25 +225,69 @@ private void parseRecentVitalResults(final Patient patient, final List<String> v
...
+                    + "(\\(.+?\\))?\\s+" // Optional capture group for the date
+                    + "(.+)\\s+" // Match the value of the vital and any information between
+                                 // the optional date and the name
+                    + "([A-Za-z0-9_,]+)"); // Match the name of the person who recorded the vital
```
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java]: You could use `\S` here, which means "non-whitespace character.

File: srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java [hunk 3]
```diff
@@ -223,25 +225,69 @@ private void parseRecentVitalResults(final Patient patient, final List<String> v
...
-            patient.getWeight().getMeasureDate(),
-            ""));
+        for(final String line: vitalResults)
+        {
+            // For each line, determine if we need to pull the vital information for that line
+            // Each entry comes with an optional date and time.
+            final Pattern pattern = Pattern.compile(
+                    "(.+:)\\s+" // Match to the first colon in order to get the name of the vital
+                    + "(\\(.+?\\))?\\s+" // Optional capture group for the date
+                    + "(.+)\\s+" // Match the value of the vital and any information between
```
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java]: "... and any _other_ information between ..." would be more accurate.

File: srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java [hunk 4]
```diff
@@ -223,25 +225,69 @@ private void parseRecentVitalResults(final Patient patient, final List<String> v
...
-        // The BMI value is the second to last token on its line
-        patient.setBmi(new RetrievedValue(
-            Double.parseDouble(bmiLineTokens.get(bmiLineTokens.size()-2)),
-            patient.getWeight().getMeasureDate(),
-            ""));
+        for(final String line: vitalResults)
+        {
+            // For each line, determine if we need to pull the vital information for that line
+            // Each entry comes with an optional date and time.
+            final Pattern pattern = Pattern.compile(
```
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java]: You can move this pattern compilation outside of the for loop for an easy performance gain.

File: srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java [hunk 5]
```diff
@@ -223,25 +225,69 @@ private void parseRecentVitalResults(final Patient patient, final List<String> v
...
+                    "(.+:)\\s+" // Match to the first colon in order to get the name of the vital
+                    + "(\\(.+?\\))?\\s+" // Optional capture group for the date
+                    + "(.+)\\s+" // Match the value of the vital and any information between
+                                 // the optional date and the name
+                    + "([A-Za-z0-9_,]+)"); // Match the name of the person who recorded the vital
+            final Matcher matcher = pattern.matcher(line);
+            matcher.matches();
```
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java]: I think you should put a conditional on this, to break early if the string is not in the expected format.

File: srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java [hunk 6]
```diff
@@ -223,25 +225,69 @@ private void parseRecentVitalResults(final Patient patient, final List<String> v
...
+            // Matcher groups begin at 1, as 0 denotes the entire pattern that was matched.
+            final String vitalName = matcher.group(1);
+            // This will be null when the date is not present.
+            final String dateString = matcher.group(2);
+            final String[] valuesArray = matcher.group(3).split(VALUES_SPLIT_REGEX);
```
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java]: Honestly, I think having `VALUES_SPLIT_REGEX` as a class constant decreases readability. It's a very simple pattern, better described by its value than a variable name.

Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java]: And why not use a Guava Splitter?

File: srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java [hunk 7]
```diff
@@ -223,25 +225,69 @@ private void parseRecentVitalResults(final Patient patient, final List<String> v
...
+                    break;
+                case "Wt.:":
+                    patient.setWeight(new RetrievedValue(
+                            // Remove any asterisks in the weight value
+                            Double.parseDouble(valuesArray[0].replace("*", "")),
+                            dateFormat.parse(dateString),
+                            WEIGHT_UNITS));
+                    break;
+                case "Body Mass Index:":
+                    /* The actual name of this vital is "Body Mass Index:" but we are splitting
```
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java]: This comment is obsolete.

File: srcalc/src/test/java/gov/va/med/srcalc/vista/RpcVistaPatientDaoTest.java [hunk 8]
```diff
@@ -370,4 +386,44 @@ public final void testInvalidDnrNotes() throws Exception
...
+    @Test
+    public final void testFullVitals() throws Exception
+    {
+        final VistaProcedureCaller caller = mockVistaProcedureCaller();
+        when(caller.doRpc(
+                RADIOLOGIST_DUZ, RemoteProcedure.GMV_LATEST_VM, String.valueOf(PATIENT_DFN)))
+            .thenReturn(FULL_VITALS);
+        final RpcVistaPatientDao dao = new RpcVistaPatientDao(caller, RADIOLOGIST_DUZ);
+        final Patient patient = dao.getPatient(PATIENT_DFN);
+        final ImmutableList<Double> expectedValues = ImmutableList.of(178.0, 71.0, 24.88);
```
Comment (dgtombs) [inline srcalc/src/test/java/gov/va/med/srcalc/vista/RpcVistaPatientDaoTest.java]: Why do you put these things into lists to check equality? Multiple calls to `assertEquals()` would be clearer and simpler, no?

File: srcalc/src/test/java/gov/va/med/srcalc/vista/RpcVistaPatientDaoTest.java [hunk 9]
```diff
@@ -370,4 +386,44 @@ public final void testInvalidDnrNotes() throws Exception
...
+    public final void testEmptyVitals()
+    {
+        final VistaProcedureCaller caller = mockVistaProcedureCaller();
+        final RpcVistaPatientDao dao = new RpcVistaPatientDao(caller, RADIOLOGIST_DUZ);
+        final Patient patient = dao.getPatient(PATIENT_DFN);
+        // Empty vitals means the vitals do not get filled with any values.
+        assertEquals(null, patient.getWeight());
+        assertEquals(null, patient.getHeight());
+        assertEquals(null, patient.getBmi());
+    }
```
Comment (dgtombs) [inline srcalc/src/test/java/gov/va/med/srcalc/vista/RpcVistaPatientDaoTest.java]: You should also add a test for unexpected data coming from VistA. Make sure your code is robust.

File: srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java [hunk 10]
```diff
@@ -223,25 +225,69 @@ private void parseRecentVitalResults(final Patient patient, final List<String> v
...
+                        // Remove any asterisks in the BMI value
+                        Double.parseDouble(valuesArray[0].replace("*", "")),
+                        patient.getWeight().getMeasureDate(),
+                        ""));
+                    break;
+            }
+        }
+    }
+    
+    private double findHeightInInches(final String[] valuesArray)
```
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java]: `valuesArray` is not a very helpful name for this parameter. Perhaps `valuesWords`?

File: srcalc/src/test/java/gov/va/med/srcalc/vista/RpcVistaPatientDaoTest.java [hunk 11]
```diff
@@ -370,4 +386,44 @@ public final void testInvalidDnrNotes() throws Exception
...
         assertEquals(0, patient.getDnrNotes().size());
     }
+    
+    @Test
+    public final void testFullVitals() throws Exception
+    {
+        final VistaProcedureCaller caller = mockVistaProcedureCaller();
+        when(caller.doRpc(
+                RADIOLOGIST_DUZ, RemoteProcedure.GMV_LATEST_VM, String.valueOf(PATIENT_DFN)))
+            .thenReturn(FULL_VITALS);
```
Comment (dgtombs) [inline srcalc/src/test/java/gov/va/med/srcalc/vista/RpcVistaPatientDaoTest.java]: The weight 6 months ago code also has no test coverage now. Perhaps you could add that here as well?

File: srcalc/src/test/java/gov/va/med/srcalc/vista/RpcVistaPatientDaoTest.java (hunk 8 above)
Comment (jamesbigmac) [inline srcalc/src/test/java/gov/va/med/srcalc/vista/RpcVistaPatientDaoTest.java]: Oh this is just unfair! You've said multiple times to put things into a list to check for equality. :(

File: srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java [hunk 12]
```diff
@@ -223,25 +224,71 @@ private void parseRecentVitalResults(final Patient patient, final List<String> v
...
+            {
+                // Invalid results were received that do not match the expected pattern.
+                break;
+            }
+            // Matcher groups begin at 1, as 0 denotes the entire pattern that was matched.
+            final String vitalName = matcher.group(1);
+            // This will be null when the date is not present.
+            final String dateString = matcher.group(2);
+            
+            final List<String> valuesWords = Splitter.on(Pattern.compile("[\\s]+")).splitToList(matcher.group(3));
```
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java]: There is `Splitter.onPattern()`.

Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/vista/RpcVistaPatientDao.java]: What is the significance of the square brackets here?

**Pull request merged**